### ISSUE TITLE
File path not properly discovered on multisite when attachment_id is missing

### ISSUE THREAD

**Issue by pippinsplugins** [easydigitaldownloads/easy-digital-downloads#3286]
During the file download process, we use various methods to try and discover the absolute path to the requested file: https://github.com/easydigitaldownloads/Easy-Digital-Downloads/blob/master/includes/process-download.php#L163

There is an issue when the following conditions are true:
- The site is a multisite install
- The download file had it's URL pasted into the URL field instead of uploaded
- The Download Method is set to `Forced`

None of these conditions is validating as true:
```php
if ( ( ! isset( $file_details['scheme'] ) || ! in_array( $file_details['scheme'], $schemes ) ) && isset( $file_details['path'] ) && file_exists( $requested_file ) ) {

	/** This is an absolute path */
	$direct    = true;
	$file_path = $requested_file;

} else if( defined( 'UPLOADS' ) && strpos( $requested_file, UPLOADS ) !== false ) {

	/**
	 * This is a local file given by URL so we need to figure out the path
	 * UPLOADS is always relative to ABSPATH
	 * site_url() is the URL to where WordPress is installed
	 */
	$file_path  = str_replace( site_url(), '', $requested_file );
	$file_path  = realpath( ABSPATH . $file_path );
	$direct     = true;

} else if( strpos( $requested_file, WP_CONTENT_URL ) !== false ) {

	/** This is a local file given by URL so we need to figure out the path */
	$file_path  = str_replace( WP_CONTENT_URL, WP_CONTENT_DIR, $requested_file );
	$file_path  = realpath( $file_path );
	$direct     = true;

} else if( strpos( $requested_file, set_url_scheme( WP_CONTENT_URL, 'https' ) ) !== false ) {

	/** This is a local file given by an HTTPS URL so we need to figure out the path */
	$file_path  = str_replace( set_url_scheme( WP_CONTENT_URL, 'https' ), WP_CONTENT_DIR, $requested_file );
	$file_path  = realpath( $file_path );
	$direct     = true;

}
```
We need to add another check for files that match the `/file` directory used by multisite.

We should also add a general `else` fallback that sets the `$file_path` to `$requested_file` and the `$direct` to `false`.

**Comment by sumobi** [comment]
see: https://easydigitaldownloads.com/support/topic/report-a-bug-in-v-2-3-2/#post-469929

**Comment by pippinsplugins** [comment]
Due to the `if/else` conditions not being met, `$file_path` is undefined.

**Comment by pippinsplugins** [comment]
This doesn't appear to affect all multi-site installs, just old ones that still use the `/files/` directory.

**Comment by pippinsplugins** [comment]
It appears this only applies to older multisite installs that still use the `blogs.dir` directory.

**Comment by pippinsplugins** [comment]
There is a work around for this (use `Redirect`) so this is not as urgent as I first thought. Punting to 2.4 so we can get 2.3.4 out.

**Comment by pippinsplugins** [comment]
This ticket appears to be an example of the problem: https://wordpress.org/support/topic/404-when-trying-to-download

**Comment by cklosowski** [comment]
This is possibly related to #1084 which we've got some work done on. Once it's merged in, I'll re-test this and see if that fixes it as well.

**Comment by pippinsplugins** [comment]
Punting this to 2.5.5

**Comment by cklosowski** [comment]
| We should also add a general else fallback that sets the $file_path to $requested_file and the $direct to false.

That was added in #4305

**Comment by cklosowski** [comment]
@pippinsplugins With the changes in #4305 will this be a necessary ticket? Is it possible to recreate the old path system for multisite? I'm failing to reproduce this one.

**Comment by pippinsplugins** [comment]
@cklosowski I'm not sure. Let's leave open but punt for now.

**Comment by pippinsplugins** [comment]
I believe this may now be resolved through #4328 but let's test to confirm.

**Comment by pippinsplugins** [comment]
I have not seen any further reports of this so I'm going to go ahead and close as fixed.
